### PR TITLE
Split ParseACM in ParseACMHeader, ValidateACMHeader, ParseACM

### PR DESCRIPTION
**Pull request by ChriMarMe**
biosACM now reads BIOS ACM Header, returns object, validates header,
reads again whole acm based on size inside acm header and parses it

### CODE REVIEW

File: pkg/test/fit.go [hunk 1]
```diff
@@ -467,14 +467,33 @@ func TestBIOSACMMatchesCPU() (bool, error) {
 func biosACM(fit []api.FitEntry) (*api.ACM, *api.Chipsets, *api.Processors, *api.TPMs, error) {
 	for _, ent := range fit {
 		if ent.Type() == api.StartUpACMod {
-			buf := make([]byte, 224*4)
+			buf1 := make([]byte, 224*4)
```
Comment (PatrickRudolph) [inline pkg/test/fit.go]: Read only 4*161

File: pkg/api/acm.go [hunk 2]
```diff
@@ -121,6 +121,37 @@ type ACM struct {
...
+
+	if err != nil {
+		return nil, fmt.Errorf("Can't read ACM Header")
+	}
+
+	return &acm, nil
+}
+
+// ValidateACMHeader
+func ValidateACMHeader(acm *ACM) (bool, error) {
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: Validate legacy header that is 4*161 in size.

File: pkg/api/acm.go [hunk 3]
```diff
@@ -121,6 +121,37 @@ type ACM struct {
...
+	buf := bytes.NewReader(data)
+	err := binary.Read(buf, binary.LittleEndian, &acm)
+
+	if err != nil {
+		return nil, fmt.Errorf("Can't read ACM Header")
+	}
+
+	return &acm, nil
+}
+
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: Add new struct "ACMLegacyHeader".

File: pkg/api/acm.go [hunk 4]
```diff
@@ -121,6 +121,37 @@ type ACM struct {
...
+func ValidateACMHeader(acm *ACM) (bool, error) {
+	if acm.ModuleType != uint16(2) {
+		return false, fmt.Errorf("BIOS ACM ModuleType is not 2, this is not specified - Intel TXT Software Development Guide, Document: 315168-013, P. 84")
+	}
+	if acm.ModuleSubType >= uint16(2) {
+		return false, fmt.Errorf("BIOS ACM ModuleSubType is greater 1, this is not specified - Intel TXT Software Development Guide, Document: 315168-013, P. 84")
+	}
+	if acm.HeaderLen < uint32(4*161) {
+		return false, fmt.Errorf("BIOS ACM HeaderLength is smaller than 4*161 Byte - Intel TXT Software Development Guide, Document: 315168-013, P. 83")
+	}
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: Check ModuleVendor

Comment (ChriMarMe) [reply pkg/api/acm.go]: done

File: pkg/api/acm.go (hunk 3 above)
Comment (ChriMarMe) [reply pkg/api/acm.go]: Did a workaround.

File: pkg/test/fit.go (hunk 1 above)
Comment (ChriMarMe) [reply pkg/test/fit.go]: Yep

File: pkg/api/acm.go [hunk 5]
```diff
@@ -117,14 +119,17 @@ type ACM struct {
 	PubKey          [256]uint8
 	PubExp          uint32
 	Signatur        [256]uint8
-	Scratch         [143]uint32
-	Info            ACMInfo
+}
+type ACM struct {
+	Header  ACMHeader
+	Scratch [143]uint32
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: variable size

File: pkg/api/acm.go [hunk 6]
```diff
@@ -136,19 +141,22 @@ func ParseACMHeader(data []byte) (*ACM, error) {
...
+func ValidateACMHeader(acmheader *ACMHeader) (bool, error) {
+	if acmheader.ModuleType != uint16(2) {
 		return false, fmt.Errorf("BIOS ACM ModuleType is not 2, this is not specified - Intel TXT Software Development Guide, Document: 315168-013, P. 84")
 	}
-	if acm.ModuleSubType >= uint16(2) {
+	if acmheader.ModuleSubType >= uint16(2) {
 		return false, fmt.Errorf("BIOS ACM ModuleSubType is greater 1, this is not specified - Intel TXT Software Development Guide, Document: 315168-013, P. 84")
 	}
-	if acm.HeaderLen < uint32(4*161) {
+	if acmheader.HeaderLen < uint32(4*161) {
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: uint32(161)

File: pkg/test/fit.go [hunk 7]
```diff
@@ -13,6 +13,7 @@ const FITSize int64 = 16 * 1024 * 1024
 const FourGiB int64 = 0x100000000
 const ResetVector = 0xFFFFFFF0
 const FITVector = 0xFFFFFFC0
+const ACMheaderLegSize = uint32(4 * 161)
```
Comment (PatrickRudolph) [inline pkg/test/fit.go]: LenSize

Comment (PatrickRudolph) [reply pkg/test/fit.go]: move to acm.go

File: pkg/api/acm.go [hunk 8]
```diff
@@ -163,13 +163,25 @@ func ValidateACMHeader(acmheader *ACMHeader) (bool, error) {
...
+	//Read only ACM Header without scratch to get scratchSize
 	buf := bytes.NewReader(data)
-	err := binary.Read(buf, binary.LittleEndian, &acm)
+	err := binary.Read(buf, binary.LittleEndian, &acmheader)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	//Generate byte slice for scratch with scratchSize from acm header
+	scratch := make([]byte, acmheader.ScratchSize)
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: acmheader.ScratchSize * 4

File: pkg/test/fit.go [hunk 9]
```diff
@@ -404,7 +403,7 @@ func TestBIOSACMSizeCorrect() (bool, error) {
 		return false, err
 	}
 
-	if acm.Header.HeaderLen%64 != 0 {
```
Comment (PatrickRudolph) [inline pkg/test/fit.go]: unrelated and wrong 64/4 = 16

File: pkg/test/fit.go [hunk 10]
```diff
@@ -403,7 +403,7 @@ func TestBIOSACMSizeCorrect() (bool, error) {
 		return false, err
 	}
 
-	if acm.Header.HeaderLen%32 != 0 {
+	if (acm.Header.HeaderLen*4)%4 == 0 {
```
Comment (PatrickRudolph) [inline pkg/test/fit.go]: condition is always true.
check is wrong, it now always returns an error.
Change is unrelated to this PR.

File: pkg/api/acm.go [hunk 11]
```diff
@@ -117,19 +120,77 @@ type ACM struct {
...
-	var acm ACM
+	var acmheader ACMHeader
 	var processors Processors
 	var chipsets Chipsets
 	var tpms TPMs
 
+	//Read only ACM Header without scratch to get scratchSize
 	buf := bytes.NewReader(data)
-	err := binary.Read(buf, binary.LittleEndian, &acm)
+	err := binary.Read(buf, binary.LittleEndian, &acmheader)
```
Comment (PatrickRudolph) [inline pkg/api/acm.go]: no need for additional local variable.
Just read acm.Header here:

```
err = binary.Read(buf, binary.LittleEndian, &acm.Header)
if err != nil {
    return nil, nil, nil, nil, err
}
```
